### PR TITLE
feat: Implement per-axis periodic mask support in cell list

### PR DESCRIPTION
**Pull request by jonkhler**
Implements the non-PBC story end-to-end on top of the new Frame/Cell hierarchy: the cell list (and the shared `basic_neighborlist` path) now honor the cell's `periodic` tuple so that `VacuumCell` produces correct edges, and the ASE loader propagates `atoms.pbc` into the kups cell so simulations driven through the application layer use the right boundary mode. Unblocks Marcel's request (issue #89) for non-PBC support.

## Three masked sites in `core/neighborlist.py`

- `_compute_distances_pbc_sq`: MIC fold gated as `jnp.where(periodic, jnp.round(deltas), 0.0)` so non-periodic axes get a zero shift.
- `_get_candidate_images`: image counts on open axes forced to 1 so the existing early-out fires for fully-open systems.
- `_cell_list_subselect`: trace-time gate on `all(periodic)`. The fully-periodic branch is byte-identical to the prior code (verified via HLO inspection); the general branch wraps only on periodic axes and routes out-of-range stencil offsets to the `cell_oob` sentinel so cross-boundary candidates produce no key matches.

## Application loader

`particles_from_ase` reads `atoms.pbc` and constructs the cell accordingly:
- all-True → `PeriodicCell(TriclinicFrame.from_matrix(...))`
- all-False → `VacuumCell(TriclinicFrame.from_matrix(...))`
- mixed → `raise NotImplementedError(...)` with a clear message

Mixed-axis (slab / wire) PBC raises today because the cell hierarchy currently exposes only the all-True/all-False literals. The neighbor-list machinery is already mask-aware via `cell.periodic` and will work transparently once a slab cell type is re-introduced (no neighborlist-side changes will be needed).

## Tests in `TestNonPeriodicNeighborList`

- Cross-boundary pair excluded under `VacuumCell`.
- In-box pair found.
- `CellListNeighborList` and `DenseNearestNeighborList` agree on a non-PBC fixture (locks in that the cell-list-specific stencil routing is consistent with the dense reference, since both go through `basic_neighborlist`).
- `PeriodicCell` path unchanged (sanity).

## Docstring updates

`core/neighborlist.py`, `core/utils/position.py`, `observables/radial_distribution_function.py`, and `potential/common/graph.py` replace "requires/uses periodic boundary conditions" with the per-axis story, since the entire stack now reads `cell.periodic` and works correctly for vacuum and bulk.

### CODE REVIEW

File: src/kups/application/utils/particles.py [hunk 1]
```diff
@@ -73,7 +79,24 @@ def particles_from_ase(
...
+        bool(atoms.pbc[0]),
+        bool(atoms.pbc[1]),
+        bool(atoms.pbc[2]),
+    )
+    cell: Cell
+    if all(pbc):
+        cell = PeriodicCell(TriclinicFrame.from_matrix(L))
+    elif not any(pbc):
+        cell = VacuumCell(TriclinicFrame.from_matrix(L))
+    else:
```
Comment (n-gao) [inline src/kups/application/utils/particles.py]: Couldn't we instantiate a Cell here?

File: src/kups/core/neighborlist.py [hunk 2]
```diff
@@ -298,8 +299,20 @@ def _cell_list_subselect(
...
+        key_positions = _wrap_coordinates(lh.data.positions)
+        query_positions = _wrap_coordinates(rh.data.positions)
+    else:
+        mask = jnp.array(periodic)
+        key_positions = jnp.where(
+            mask, _wrap_coordinates(lh.data.positions), lh.data.positions
+        )
+        query_positions = jnp.where(
+            mask, _wrap_coordinates(rh.data.positions), rh.data.positions
+        )
```
Comment (n-gao) [inline src/kups/core/neighborlist.py]: Shouldn't cell's .wrap take care of this?

File: src/kups/core/neighborlist.py [hunk 3]
```diff
@@ -325,17 +338,31 @@ def _cell_list_subselect(
...
+        # offsets on open axes route to cell_oob so they produce no key matches
+        # (cross-boundary candidates are excluded).
+        mask = jnp.array(periodic)
+        shifted = jnp.where(mask, _wrap_coordinates(raw_shifted), raw_shifted)
+        in_box = jnp.all((shifted >= 0) & (shifted < 1), axis=-1)
+        hashes = (
+            _cell_hash(shifted, num_cells[query_system])
+            + rh_system_ids[query_original.indices] * max_num_cells.size
+        )
+        query_neighborhood_hashes = jnp.where(in_box, hashes, cell_oob)
```
Comment (n-gao) [inline src/kups/core/neighborlist.py]: Again cell.wrap should handle the logic of periodic. Ideally, we do not need to concern ourselfs with .periodic outside as much as possible

File: src/kups/core/neighborlist.py [hunk 4]
```diff
@@ -553,7 +583,8 @@ def _compute_distances_pbc_sq(
         - rh.data.positions[candidates.rhs.indices]
     )
     if shifts is None:
-        shifts = jnp.round(deltas)
+        periodic = jnp.array(systems.data.cell.periodic)
+        shifts = jnp.where(periodic, jnp.round(deltas), 0.0)
```
Comment (n-gao) [inline src/kups/core/neighborlist.py]: Should be a Cell method

File: test/core/test_neighborlist.py [hunk 5]
```diff
@@ -1762,3 +1768,106 @@ def test_random_large(self):
...
+        avg_image_candidates=FixedCapacity(max(n * n, 8)),
+    )
+    result = jax.jit(as_result_function(nl))(
+        lh=lh, rh=None, systems=systems, cutoffs=cutoffs, rh_index_remap=None
+    )
+    result.raise_assertion()
+    return result.value
+
+
+class TestNonPeriodicNeighborList:
```
Comment (n-gao) [inline test/core/test_neighborlist.py]: Can we directly test all variations of periodicities? vaccuum, 1D, 2D, 3d?

File: src/kups/core/utils/position.py
Comment (n-gao) [inline src/kups/core/utils/position.py:1]: This file needs to be updated as well to respect periodicity, right?

**Review by n-gao — CHANGES_REQUESTED** [review]
The general changes are sound to me! However, I would avoid the user-side code (in this case everything outside of cell.py) dealing with the .periodic tuple as much as possible and would instead encapsulate most of the handling to the existing .wrap method and new methods were appropriate. This would help us to reduce errors in the future.

File: src/kups/core/cell.py [hunk 6]
```diff
@@ -465,9 +495,7 @@ class PeriodicCell(Cell[Periodic3D]):
     Ewald summation requires this cell type.
     """
 
-    @property
-    def periodic(self) -> Periodic3D:
-        return (True, True, True)
+    periodic: Periodic3D = field(default=(True, True, True), static=True)
```
Comment (n-gao) [inline src/kups/core/cell.py]: init=False

File: src/kups/core/cell.py [hunk 7]
```diff
@@ -481,9 +509,7 @@ class VacuumCell(Cell[Vacuum]):
     neighbor-list machinery needs a spatial-partitioning hint.
     """
 
-    @property
-    def periodic(self) -> Vacuum:
-        return (False, False, False)
+    periodic: Vacuum = field(default=(False, False, False), static=True)
```
Comment (n-gao) [inline src/kups/core/cell.py]: init=False

File: src/kups/application/utils/particles.py [hunk 8]
```diff
@@ -73,8 +79,22 @@ def particles_from_ase(
...
+    cell: Cell
+    match pbc:
+        case (True, True, True):
+            cell = PeriodicCell(frame)
+        case (False, False, False):
+            cell = VacuumCell(frame)
+        case (True, True, False) | (True, False, True) | (False, True, True):
+            cell = Cell(frame, periodic=pbc)
+        case (True, False, False) | (False, True, False) | (False, False, True):
+            cell = Cell(frame, periodic=pbc)
```
Comment (n-gao) [inline src/kups/application/utils/particles.py]: this could be a staticmethod of Cell?

**Review by n-gao — APPROVED** [review]
LGTM, we could introduce a staticmethod for constructing the cell and set init=False for the Vacuum and PeriodicCell.
